File: tests/test_api.js
import request from 'supertest';
import { createApp } from '../backend/server.js';

describe('REST API', () => {
  let app;
  let marketDataService;

  beforeAll(() => {
    ({ app, marketDataService } = createApp({ marketOptions: { intervalMs: 3600000 } }));
    marketDataService.tick();
  });

  afterAll(() => {
    marketDataService.shutdown();
  });

  test('GET /api/assets returns synthetic instruments', async () => {
    const response = await request(app).get('/api/assets');
    expect(response.status).toBe(200);
    expect(Array.isArray(response.body.assets)).toBe(true);
    expect(response.body.assets.length).toBeGreaterThan(0);
  });

  test('GET /api/history/:symbol returns candle history', async () => {
    const symbol = marketDataService.assets[0].symbol;
    const response = await request(app).get(`/api/history/${symbol}`);
    expect(response.status).toBe(200);
    expect(response.body.symbol).toBe(symbol);
    expect(Array.isArray(response.body.candles)).toBe(true);
    expect(response.body.candles.length).toBeGreaterThan(0);
  });

  test('POST /api/orders places manual trade', async () => {
    const symbol = marketDataService.assets[0].symbol;
    const response = await request(app).post('/api/orders').send({ symbol, quantity: 1 });
    expect(response.status).toBe(201);
    expect(response.body.symbol).toBe(symbol);

    const portfolio = await request(app).get('/api/portfolio');
    expect(portfolio.status).toBe(200);
    const position = portfolio.body.positions.find((p) => p.symbol === symbol);
    expect(position).toBeDefined();
    expect(position.quantity).toBeGreaterThan(0);
  });
});

File: tests/test_api.py
"""Integration tests for the FastAPI trading backend."""

from __future__ import annotations

from fastapi.testclient import TestClient
import pytest

from pybackend.server import app, market_data_service, trading_mode_service


@pytest.fixture(scope="module")
def client():
    with TestClient(app) as test_client:
        yield test_client


def ensure_paper_state():
    trading_mode_service.set_mode("paper")
    market_data_service.tick()


def test_get_assets(client):
    ensure_paper_state()
    response = client.get("/api/assets")
    assert response.status_code == 200
    payload = response.json()
    assert "assets" in payload
    assert isinstance(payload["assets"], list)
    assert len(payload["assets"]) > 0


def test_get_history(client):
    ensure_paper_state()
    symbol = market_data_service.assets[0]["symbol"]
    response = client.get(f"/api/history/{symbol}")
    assert response.status_code == 200
    body = response.json()
    assert body["symbol"] == symbol
    assert isinstance(body["candles"], list)
    assert len(body["candles"]) > 0


def test_post_orders_updates_portfolio(client):
    ensure_paper_state()
    symbol = market_data_service.assets[0]["symbol"]
    order_response = client.post("/api/orders", json={"symbol": symbol, "quantity": 1})
    assert order_response.status_code == 201
    trade = order_response.json()
    assert trade["symbol"] == symbol

    portfolio_response = client.get("/api/portfolio")
    assert portfolio_response.status_code == 200
    portfolio = portfolio_response.json()
    positions = portfolio.get("positions", [])
    assert any(position["symbol"] == symbol and position["quantity"] > 0 for position in positions)
